### ISSUE TITLE
Flaky bug in ActionController::Parameters#permit!

### ISSUE THREAD

**Issue by olefriis** [rails/rails#45220]
### Steps to reproduce
In GitHub, after upgrading to Rails 7.1 de12c18a99, a few tests have become flaky. They call endpoints with `params` that contain lists with hashes.

When the tests fail, the failure is
```
ActiveSupport::DeprecationException: DEPRECATION WARNING: Comparing equality between `ActionController::Parameters` and a `Hash` is deprecated and will be removed in Rails 7.2. Please only do comparisons between instances of `ActionController::Parameters`. If you need to compare to a hash, first convert it using `ActionController::Parameters#new`. To disable the deprecated behaviour set `Rails.application.config.action_controller.allow_deprecated_parameters_hash_equality = false`. (called from [...] at [...])
```

## Reproduction
The error is reproducible locally by running one of the tests over, over, and over again (e.g. `while bin/rails test <path-to-test>; do echo next; done`). We instrumented `ActionController::Parameters#==` (in `actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb`) with a bit of backtrace stuff so it now looks like this:
```ruby
    def ==(other)
      if other.respond_to?(:permitted?)
        permitted? == other.permitted? && parameters == other.parameters
      else
        begin
          raise "We're here!"
        rescue => e
          puts e.backtrace
        end
        puts "Other class: #{other.class}"
        if Hash === other
          puts 'It is a Hash!'
        end
        if self.class.allow_deprecated_parameters_hash_equality && Hash === other
          ActiveSupport::Deprecation.warn <<-WARNING.squish
            Comparing equality between `ActionController::Parameters` and a
            `Hash` is deprecated and will be removed in Rails 7.2. Please only do
            comparisons between instances of `ActionController::Parameters`. If
            you need to compare to a hash, first convert it using
            `ActionController::Parameters#new`.
            To disable the deprecated behaviour set
            `Rails.application.config.action_controller.allow_deprecated_parameters_hash_equality = false`.
          WARNING
          @parameters == other
        else
          super
        end
      end
    end
```
...and got this nice, little stack trace:
```
Run options: --seed 55159

# Running:

......................................................................................We're here!
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:287:in `=='
/workspaces/github/vendor/ruby/4a0738122107d0d7af669feed62452ce64568692/lib/ruby/3.1.0/set.rb:404:in `eql?'
/workspaces/github/vendor/ruby/4a0738122107d0d7af669feed62452ce64568692/lib/ruby/3.1.0/set.rb:404:in `eql?'
/workspaces/github/vendor/ruby/4a0738122107d0d7af669feed62452ce64568692/lib/ruby/3.1.0/set.rb:404:in `include?'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:1004:in `convert_value_to_parameters'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:996:in `convert_hashes_to_parameters'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:407:in `block in each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:406:in `each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:406:in `each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:464:in `permit!'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:466:in `block (2 levels) in permit!'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:465:in `each'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:465:in `block in permit!'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:407:in `block in each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:406:in `each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:406:in `each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:464:in `permit!'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:466:in `block (2 levels) in permit!'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:465:in `each'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:465:in `block in permit!'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:407:in `block in each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:406:in `each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:406:in `each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:464:in `permit!'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:466:in `block (2 levels) in permit!'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:465:in `each'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:465:in `block in permit!'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:407:in `block in each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:406:in `each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:406:in `each_pair'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/action_controller/metal/strong_parameters.rb:464:in `permit!'
[...here is our controller code calling permit! on an ActiveSupport::Parameters instance...]
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/activesupport-7.1.0.alpha.de12c18a99/lib/active_support/callbacks.rb:400:in `block in make_lambda'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/activesupport-7.1.0.alpha.de12c18a99/lib/active_support/callbacks.rb:180:in `block (2 levels) in halting_and_conditional'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/actionpack-7.1.0.alpha.de12c18a99/lib/abstract_controller/callbacks.rb:34:in `block (2 levels) in <module:Callbacks>'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/activesupport-7.1.0.alpha.de12c18a99/lib/active_support/callbacks.rb:181:in `block in halting_and_conditional'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/activesupport-7.1.0.alpha.de12c18a99/lib/active_support/callbacks.rb:595:in `block in invoke_before'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/activesupport-7.1.0.alpha.de12c18a99/lib/active_support/callbacks.rb:595:in `each'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/activesupport-7.1.0.alpha.de12c18a99/lib/active_support/callbacks.rb:595:in `invoke_before'
/workspaces/github/vendor/gems/3.1.2/ruby/3.1.0/gems/activesupport-7.1.0.alpha.de12c18a99/lib/active_support/callbacks.rb:116:in `block in run_callbacks'
[...more lines probably irrelevant for this issue...]
[...our test, POSTing to the endpoint...]
[...and MiniTest stack trace etc...]
Other class: ActiveSupport::HashWithIndifferentAccess
It is a Hash!
```

## Analysis
So sometimes, the implementation of `permit!` compares an `ActiveSupport::HashWithIndifferentAccess` with an `ActionController::Parameters` all by itself.

As the stack trace suggests, it happens here:
```ruby
      def convert_value_to_parameters(value)
        case value
        when Array
          return value if converted_arrays.member?(value)
          converted = value.map { |_| convert_value_to_parameters(_) }
          converted_arrays << converted.dup
          converted
        when Hash
          self.class.new(value, @logging_context)
        else
          value
        end
      end
```
(The `return value if converted_arrays.member?(value)` line.)

Most of the time, the `converted_arrays` object, which is a `Set`, will have no hash collisions with `value`, but sometimes it will, and then it needs to compare existing entries in the `Set` with `value`. And since `value` has not been converted to an `ActionController::Parameters` object, we get the error.

We have changed the `convert_value_to_parameters` method to provoke this error deterministically by making it a bit uglier and much less performant, so we compare all existing elements with `value` 😃
```ruby
      def convert_value_to_parameters(value)
        case value
        when Array
          match = false
          converted_arrays.each do |element|
            match ||= element == value
          end

          return value if match
          converted = value.map { |_| convert_value_to_parameters(_) }
          converted_arrays << converted.dup
          converted
        when Hash
          self.class.new(value, @logging_context)
        else
          value
        end
      end
```

## Minimal (?) test
With the changes to `convert_value_to_parameters` above, this test triggers the issue consistently:
```ruby
  test "strong parameters may be buggy" do
    ActionController::Parameters.new(
      "iterations": [
        {
          "title": "Iteration 1",
        },
      ],
      "completedIterations": [
        {
          "title": "Iteration 2",
        },
      ],
    ).permit!
  end
```
Maybe it can be made even smaller, but we're not sure.

## Possible work-around
Changing the above test a bit so that it does part of the work of `convert_value_to_parameters` can make it pass:
```ruby
  test "strong parameters may be buggy" do
    ActionController::Parameters.new(
      "iterations": [
        ActionController::Parameters.new(
          "title": "Iteration 1",
        ),
      ],
      "completedIterations": [
        ActionController::Parameters.new(
          "title": "Iteration 2",
        ),
      ],
    ).permit!
  end
```
...but this just seems wrong...? I believe the failing test should not fail, or...?

## Proposed Fix
This should fix the issue:
```ruby
      def convert_value_to_parameters(value)
        case value
        when Array
          converted = value.map { |_| convert_value_to_parameters(_) }
          return value if converted_arrays.member?(converted)
          converted_arrays << converted.dup
          converted
        when Hash
          self.class.new(value, @logging_context)
        else
          value
        end
      end
```
Now we convert the value before trying to check for membership in `converted_arrays`. This is less performant, but is maybe not too bad, since `convert_value_to_parameters` will just return the given value if it is already converted.

Maybe we can just remove the cache:
```ruby
      def convert_value_to_parameters(value)
        case value
        when Array
          value.map { |_| convert_value_to_parameters(_) }
        when Hash
          self.class.new(value, @logging_context)
        else
          value
        end
      end
```

Also, writing a test that shows a _lack_ of flakiness also needs to be done...

## Further Context
Interestingly, there was a fix in 2021 in the same method: https://github.com/rails/rails/issues/43681

But this seems to be a different issue that is related to the phasing-out of comparison between `ActionController::Parameters` and `Hash`: https://github.com/rails/rails/commit/8af86c997c7bfbb05afa7dbde8847cf9b55a7b47

(Sorry, I haven't created a reproduction script according to the template. I can do that, though the above test should be easy to get working. Let me know if I should expand on this.)

### Expected behavior
`ActionController::Parameters#permit!` shouldn't be flaky 😄 

### Actual behavior
`ActionController::Parameters#permit!` is flaky 😄 

### System configuration
**Rails version**: Rails 7.1.0.alpha.de12c18a99

**Ruby version**: ruby 3.1.2p20 (2022-04-12 revision 4a07381221) [x86_64-linux]

**Comment by jhawthorn** [comment]
This may be the same as #44940

**Comment by olefriis** [comment]
@jhawthorn At least they are _very_ tightly related. While this issue is a problem with checking membership in the `Set` where the checked value has not yet been converted to a `Parameters` object, the other issue is about adding the converted `Parameters` object to the `Set`. And while this issue is about flaky behaviour, the other is deterministic.

Amazing. I'm still trying to wrap my head around all of it.

But your fix works for both, since `Set` uses `eql?`, so both `Set#member?` (alias for `Set#include?` which is what's in the stacktrace above) and `Set#<<` (alias for `Set#add`) will now have a `eql?` that does _not_ provoke the deprecation message.

I'll be running our flaky tests for a couple of hours with the changes in your fix, just to gain more confidence that it works, but I strongly believe this fixes it.

Thank you SO much! ❤️

**Comment by jhawthorn** [comment]
Fixed by #45221
